File: python/3sumclosest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        curMin = float('inf')
        cur = nums[0] + nums[1] + nums[2]
        for i in xrange(len(nums)-2):
            j = i+1
            k = len(nums)-1
            while j<k:
                tmp = nums[i] + nums[j] + nums[k]
                if tmp == target:
                    return target
                elif tmp < target:
                    j += 1
                else:
                    k -=1
                if abs(target-tmp) < curMin:
                    curMin = abs(target-tmp)
                    cur = tmp
        return cur

File: cpp/3sumclosest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int minVal = INT_MAX;
        int minGap = INT_MAX;
        sort(nums.begin(), nums.end(), [](const int &i, const int &j){return i<j;});
        for (int i = 0; i < nums.size()-2; i++) {
            int j = i+1, k = nums.size()-1;
            while (j < k) {
                int tmp = nums[i] + nums[j] + nums[k];
                if (tmp == target) return target;
                if (abs(tmp-target) < minGap) {
                    minGap = abs(tmp-target);
                    minVal = tmp;
                }
                if (tmp < target) j++;
                else k--;
            }
        }
        return minVal;
        
    }
};
